pandas/core/groupby/groupby.py

```diff
@@ -4990,25 +4997,67 @@ def shift(
         else:
             axis = 0
 
-        if freq is not None or axis != 0:
-            f = lambda x: x.shift(periods, freq, axis, fill_value)
-            return self._python_apply_general(f, self._selected_obj, is_transform=True)
+        if is_list_like(periods):
+            periods = cast(list, periods)
+            if axis == 1:
+                raise ValueError(
+                    "If `periods` contains multiple shifts, `axis` cannot be 1."
+                )
+            if len(periods) == 0:
+                raise ValueError("If `periods` is an iterable, it cannot be empty.")
+            from pandas.core.reshape.concat import concat
 
-        if fill_value is lib.no_default:
-            fill_value = None
-        ids, _, ngroups = self.grouper.group_info
-        res_indexer = np.zeros(len(ids), dtype=np.int64)
+            add_suffix = True
+        else:
+            if not isinstance(periods, int):
+                raise TypeError(
+                    f"Periods must be integer, but {periods} is {type(periods)}."
+                )
+            periods = [periods]
+            add_suffix = False
+
+        shifted_dataframes = []
+        for period in periods:
+            if not isinstance(period, int):
+                raise TypeError(
+                    f"Periods must be integer, but {period} is {type(period)}."
+                )
+            if freq is not None or axis != 0:
+                f = lambda x: x.shift(
+                    period, freq, axis, fill_value  # pylint: disable=cell-var-from-loop
+                )
+                shifted = self._python_apply_general(
+                    f, self._selected_obj, is_transform=True
+                )
+            else:
+                if fill_value is lib.no_default:
+                    fill_value = None
+                ids, _, ngroups = self.grouper.group_info
+                res_indexer = np.zeros(len(ids), dtype=np.int64)
 
-        libgroupby.group_shift_indexer(res_indexer, ids, ngroups, periods)
+                libgroupby.group_shift_indexer(res_indexer, ids, ngroups, period)
 
-        obj = self._obj_with_exclusions
+                obj = self._obj_with_exclusions
 
-        res = obj._reindex_with_indexers(
-            {self.axis: (obj.axes[self.axis], res_indexer)},
-            fill_value=fill_value,
-            allow_dups=True,
+                shifted = obj._reindex_with_indexers(
+                    {self.axis: (obj.axes[self.axis], res_indexer)},
+                    fill_value=fill_value,
+                    allow_dups=True,
+                )
+
+            if add_suffix:
+                if len(shifted.shape) == 1:
```

Comment:
Instead of this check, can you do `isinstance(shifted, Series)` throughout? I think that's what you mean to check here.

Reply:
👍